File: lab3/main.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <cmath>

class GaussBlur {
public:
    int n;
    cv::Mat gauss;

    GaussBlur(double eps, int n_) {
        int a = n_ / 2;
        int b = n_ / 2;

        n = n_;

        gauss = cv::Mat(n_, n_, CV_64F);

        double sum = 0.0;
        for (int y = 0; y < n_; y++) {
            for (int x = 0; x < n_; x++) {
                double val = (1.0 / (2 * M_PI * eps * eps)) *
                             std::exp(-((x - a) * (x - a) + (y - b) * (y - b)) / (2 * eps * eps));
                gauss.at<double>(y, x) = val;
                sum += val;
            }
        }
        gauss /= sum;
    }

    cv::Vec3d apply_kernel(const cv::Mat &matr) {
        cv::Vec3d res(0, 0, 0);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                cv::Vec3b pix = matr.at<cv::Vec3b>(i, j);
                double k = gauss.at<double>(i, j);
                res[0] += pix[0] * k;
                res[1] += pix[1] * k;
                res[2] += pix[2] * k;
            }
        }
        return res;
    }

    cv::Mat blur(const cv::Mat img) {
        cv::Mat padded;
        int border = n / 2;

        cv::copyMakeBorder(img, padded, border, border, border, border, cv::BORDER_REPLICATE);

        cv::Mat new_img = cv::Mat::zeros(img.size(), img.type());

        for (int y = 0; y < img.rows; y++) {
            for (int x = 0; x < img.cols; x++) {
                cv::Rect rect(x, y, n, n);
                cv::Mat pad_rect = padded(rect);
                cv::Vec3d val = apply_kernel(pad_rect);
                new_img.at<cv::Vec3b>(y, x) = cv::Vec3b(
                    cv::saturate_cast<uchar>(val[0]),
                    cv::saturate_cast<uchar>(val[1]),
                    cv::saturate_cast<uchar>(val[2])
                );
            }
        }

        return new_img;
    }

    cv::Mat getKernel() {
        return gauss;
    }
};

int main(int argc, char *argv[]) {
    int n = 3;
    if (argc > 1) { n = std::atoi(argv[1]); }
    double eps = 0.84089642;
    if (argc > 2) { eps = atof(argv[2]); }

    GaussBlur gauss(eps, n);

    cv::Mat img = cv::imread("./img.png");

    cv::imshow("Window_orig", img);

    cv::Mat my_blur = gauss.blur(img);
    cv::imshow("Window blur", my_blur);

    cv::Mat cv2_blur;
    cv::GaussianBlur(img, cv2_blur, cv::Size(n, n), eps);
    cv::imshow("Window CV2", cv2_blur);

    cv::waitKey(0);
    cv::destroyAllWindows();
    return 0;
}

File: lab3/main.py
import numpy as np
import cv2

class GaussBlur:
    def __init__(self, eps, n):
        self.n = n
        a = self.n//2
        b = self.n//2
        self.gauss = []
        for y in range(n):
            self.gauss.append([])
            for x in range(n):
                self.gauss[-1].append(1/(2*np.pi*eps*eps)*np.e**(-((x-a)**2+(y-b)**2)/(2*eps*eps)))
        self.gauss = np.array(self.gauss)
        self.gauss /= np.sum(self.gauss)

    def apply_kernel(self, matr):
        res = np.zeros(3)
        for i in range(len(matr)):
            for j in range(len(matr[i])):
                res += matr[i][j] * self.gauss[i][j]
        return res

    def blur(self, img):
        img_temp = img

        new_img = np.zeros_like(img)

        for _ in range(self.n//2):
            img_temp = np.insert(img_temp, 0, img_temp[0], axis=0)
            img_temp = np.insert(img_temp, -1, img_temp[-1], axis=0)
        for _ in range(self.n//2):
            img_temp = np.insert(img_temp, 0, img_temp[:,0], axis=1)
            img_temp = np.insert(img_temp, -1, img_temp[:,-1], axis=1)

        for y in range(len(img)):
            ty = y+self.n//2
            for x in range(len(img[y])):
                tx = x+self.n//2
                new_img[y, x] = self.apply_kernel(img_temp[ty-(self.n-1)//2:ty+(self.n-1)//2+1, tx-(self.n-1)//2:tx+(self.n-1)//2+1])

        return new_img

gauss = GaussBlur(15, 3)

img = cv2.imread("media/img.png")
cv2.imshow("Window_orig", img)
cv2.imshow("Window blur", gauss.blur(img))

cv2_img = cv2.filter2D(src=img, ddepth=-1, kernel=gauss.gauss)
cv2.imshow("Window CV2", cv2_img)

cv2.waitKey(0)
cv2.destroyAllWindows()